pkg/render/gateway_api.go

```diff
@@ -0,0 +1,507 @@
+	}
+	for _, crd := range resources.envoyCRDs {
+		gatewayAPICRDs = append(gatewayAPICRDs, crd.DeepCopyObject().(client.Object))
+	}
+
+	return gatewayAPICRDs
+}
+
+type GatewayAPIImplementationConfig struct {
+	Installation *operatorv1.InstallationSpec
+	GatewayAPI   *operatorv1.GatewayAPI
+	PullSecrets  []*corev1.Secret
+}
+
+type gatewayAPIImplementationComponent struct {
+	cfg                 *GatewayAPIImplementationConfig
+	envoyGatewayImage   string
+	envoyProxyImage     string
+	envoyRatelimitImage string
+}
+
+func GatewayAPIImplementationComponent(cfg *GatewayAPIImplementationConfig) Component {
+	return &gatewayAPIImplementationComponent{
+		cfg: cfg,
+	}
+}
+
+func (pr *gatewayAPIImplementationComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := pr.cfg.Installation.Registry
+	path := pr.cfg.Installation.ImagePath
+	prefix := pr.cfg.Installation.ImagePrefix
+
+	var err error
+	pr.envoyGatewayImage, err = components.GetReference(components.ComponentGatewayAPIEnvoyGateway, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	pr.envoyProxyImage, err = components.GetReference(components.ComponentGatewayAPIEnvoyProxy, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	pr.envoyRatelimitImage, err = components.GetReference(components.ComponentGatewayAPIEnvoyRatelimit, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (pr *gatewayAPIImplementationComponent) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+func (pr *gatewayAPIImplementationComponent) Ready() bool {
+	return true
+}
+
+func (pr *gatewayAPIImplementationComponent) Objects() ([]client.Object, []client.Object) {
+	resources := GatewayAPIResources()
+
+	// List all the non-CRD resources, read from YAML, that we can apply without any tweaking.
+	objs := []client.Object{
+		resources.namespace,
+		resources.controllerServiceAccount,
+		resources.clusterRole,
+		resources.clusterRoleBinding,
+		resources.role,
+		resources.roleBinding,
+		resources.leaderElectionRole,
+		resources.leaderElectionRoleBinding,
+		resources.controllerService,
+		resources.certgenServiceAccount,
+		resources.certgenRole,
+		resources.certgenRoleBinding,
+		pr.envoyProxyConfig(),
+	}
+
+	// Deep-copy all of those, so as not to inadvertently modify the cache inside
+	// `GatewayAPIResourcesGetter`.
+	for i := range objs {
+		objs[i] = objs[i].DeepCopyObject().(client.Object)
+	}
+
+	// Substitute possibly modified image names into the envoy-gateway-config.
+	envoyGatewayConfig := resources.envoyGatewayConfig.DeepCopyObject().(*envoyapi.EnvoyGateway)
+	envoyGatewayConfig.Provider.Kubernetes.RateLimitDeployment.Container.Image = &pr.envoyRatelimitImage
+	envoyGatewayConfig.Provider.Kubernetes.ShutdownManager.Image = &pr.envoyGatewayImage
+
+	// Add pull secrets into the envoy-gateway-config, insofar as the API allows this.  (In
+	// other words, currently only for the ratelimit image.  So effectively the gateway image
```

Comment:
Is this comment correct? It looks like you are able to set the pull secret on the deployment  in line 447.

Reply:
Ah yes, you are right.  I thought the envoy coding and config was still missing a way to set pull secrets for the "ShutdownManager" image - because the config only allows setting an image - but in fact pull secrets are set at the pod level and so the handling at
```
	envoyGatewayConfig.Provider.Kubernetes.RateLimitDeployment.Pod.ImagePullSecrets = secret.GetReferenceList(pr.cfg.PullSecrets)
```
actually covers the "ShutdownManager" image as well as the "RateLimit" image.